### PR TITLE
[IOPID-638] : Starting with unit tests

### PR DESCRIPTION
**Pull request by claudiodall**
## Short description

Some unit tests have been added using Jest for several pages in order to provide adequate coverage.

## How to test

To test, run yarn test to execute the tests, and yarn test --coverage to see the coverage.

### CODE REVIEW

File: .env.local [hunk 1]
```diff
@@ -5,3 +5,6 @@ NEXT_PUBLIC_URL_IO=https://io.italia.it
 NEXT_PUBLIC_JWT_SPID_LEVEL_VALUE_L1=https://www.spid.gov.it/SpidL1
 NEXT_PUBLIC_JWT_SPID_LEVEL_VALUE_L2=https://www.spid.gov.it/SpidL2
 NEXT_PUBLIC_JWT_SPID_LEVEL_VALUE_L3=https://www.spid.gov.it/SpidL3
+
+# MOCK API FOR TEST ENVIRONMENT
+NEXT_PUBLIC_API_MOCK_WEB_PROFILE=false
```
Comment (shadowsheep1) [inline .env.local:10]: Is this related with tests in this PR?

File: .env.production [hunk 2]
```diff
@@ -4,4 +4,7 @@ NEXT_PUBLIC_URL_IO=https://io.italia.it
 # TOKEN SPID LEVEL VALUE
 NEXT_PUBLIC_JWT_SPID_LEVEL_VALUE_L1=https://www.spid.gov.it/SpidL1
 NEXT_PUBLIC_JWT_SPID_LEVEL_VALUE_L2=https://www.spid.gov.it/SpidL2
-NEXT_PUBLIC_JWT_SPID_LEVEL_VALUE_L3=https://www.spid.gov.it/SpidL3
\ No newline at end of file
+NEXT_PUBLIC_JWT_SPID_LEVEL_VALUE_L3=https://www.spid.gov.it/SpidL3
+
+# MOCK API FOR TEST ENVIRONMENT
+NEXT_PUBLIC_API_MOCK_WEB_PROFILE=false
```
Comment (shadowsheep1) [inline .env.production:10]: As above

File: .env.test.local [hunk 3]
```diff
@@ -1 +1,3 @@
-NEXT_PUBLIC_URL_IO=https://io.italia.it
\ No newline at end of file
+NEXT_PUBLIC_URL_IO=https://io.italia.it
+# MOCK API FOR TEST ENVIRONMENT
+NEXT_PUBLIC_API_MOCK_WEB_PROFILE=true
```
Comment (shadowsheep1) [inline .env.test.local:3]: As above

File: src/app/[locale]/(pages)/accedi/__test__/page.test.tsx [hunk 4]
```diff
@@ -1,25 +1,13 @@
-import { render, screen } from '@testing-library/react';
-import { NextIntlProvider } from 'next-intl';
+/* eslint-disable functional/immutable-data */
```
Comment (shadowsheep1) [inline src/app/[locale]/(pages)/accedi/__test__/page.test.tsx:1]: Is this needed?

File: src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+}));
+
+// Mock the behavior of useSearchParams
+const mockSearchParams = {
+  get: jest.fn(),
+};
+
+(useSearchParams as jest.Mock).mockReturnValue(mockSearchParams);
+
+describe('test suite for Accedi Errore ', () => {
```
Comment (shadowsheep1) [inline src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx:20]: Here we are mixing English and Italian. Is this wanted? I mean in the test description, not in the expected values.

File: src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+      'Per accedere al portale, devi acconsentire all’invio di alcuni dati. Entra di nuovo e dai il tuo consenso.'
+    );
+    expect(errorSummary).toBeInTheDocument();
+  });
+
+  test('should render "Spid Revoked" for errorCode 23', async () => {
+    mockSearchParams.get.mockReturnValue('23');
+    await renderWithProviders(<LoginErrorPage />);
+    const errorSummary = screen.getByText(
+      'La tua identità SPID risulta sospesa o revocata. Per riattivare il tuo account, contatta il tuo fornitore di identità SPID.'
```
Comment (shadowsheep1) [inline src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx:70]: If we already have translations for the UI component texts, are we able to use them so that if they change we don't need to modify our tests?

File: src/app/[locale]/(pages)/accedi/errore/page.tsx [hunk 7]
```diff
@@ -96,7 +96,7 @@ const LoginErrorPage = () => {
                 onClick={() => pushWithLocale(process.env.NEXT_PUBLIC_URL_IO || '')}
                 variant="contained"
               >
-                Torna alla home
+                {t('common.backtohome')}
```
Comment (shadowsheep1) [inline src/app/[locale]/(pages)/accedi/errore/page.tsx:99]: This is not related to this PR I guess, but I did understand why you did it ;)

File: src/app/[locale]/_utils/__tests__/common.test.tsx [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import { storageLocaleOps } from './../storage';
+import { addSpacesEvery3Chars, localeFromStorage, isBrowser } from '../common';
+import * as common from '../common';
+
+afterEach(() => {
+  jest.clearAllMocks();
+});
+
+describe('Utilities', () => {
+  describe('addSpacesEvery3Chars', () => {
```
Comment (shadowsheep1) [inline src/app/[locale]/_utils/__tests__/common.test.tsx:10]: Should we improve those descriptions? I mean, can we be more expressive?

File: src/app/[locale]/_utils/__tests__/jwt.test.tsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
+import { JWTUser } from '../../_model/JWTUser';
+import { User } from '../../_model/User';
+import { extractToken, parseJwt, userFromJwtToken } from '../jwt';
+
+describe('Utilities', () => {
+  describe('extractToken', () => {
+    test('extracts token from URL', () => {
```
Comment (shadowsheep1) [inline src/app/[locale]/_utils/__tests__/jwt.test.tsx:7]: Should we be more expressive in those descriptions?

File: src/app/[locale]/_utils/__tests__/jwt.test.tsx [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+      const result = extractToken();
+      expect(result).toBe('mytesttoken');
+
+      Object.defineProperty(window, 'location', { value: originalLocation });
+    });
+
+  });
+
+  describe('parseJwt', () => {
+    test('parses JWT token', () => {
```
Comment (shadowsheep1) [inline src/app/[locale]/_utils/__tests__/jwt.test.tsx:22]: As above

File: src/app/[locale]/_utils/__tests__/storage-utils.test.tsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+  describe('Storage Functions', () => {
+    const key = 'testKey';
+    const value = 'testValue';
+  
+    beforeEach(() => {
+      localStorage.clear();
+      sessionStorage.clear();
+    });
+  
+    test('storageDelete deletes the correct key', () => {
```
Comment (shadowsheep1) [inline src/app/[locale]/_utils/__tests__/storage-utils.test.tsx:18]: As above.  For example: "storageDelete should delete the item corresponding to the passed key". Just to be consistent with jest wording. Even without "should" (if we want to follow the [spotify philosophy](https://github.com/spotify/should-up), but then we must be consistent in all our description. To me, I like "should". BTW, I think that having more verbose description helps. What do you think about that?

**Review by shadowsheep1 — CHANGES_REQUESTED** [review]

File: .env.local (hunk 1 above)
Comment (claudiodall) [reply .env.local:10]: seems to be something related to something else, sorry, let's remove it

File: .env.test.local (hunk 3 above)
Comment (claudiodall) [reply .env.test.local:3]: seems to be something related to something else, sorry, let's remove it

File: .env.production (hunk 2 above)
Comment (claudiodall) [reply .env.production:10]: seems to be something related to something else, sorry, let's remove it

File: src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx (hunk 5 above)
Comment (claudiodall) [reply src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx:20]: It's a mistake, let's fix it

File: src/app/[locale]/(pages)/accedi/__test__/page.test.tsx (hunk 4 above)
Comment (claudiodall) [reply src/app/[locale]/(pages)/accedi/__test__/page.test.tsx:1]: Since we were getting errors while creating the tests with next-intl, we created a utils according to docs, have a look at -> '/app/[locale]/_utils/test-utils'

Comment (shadowsheep1) [reply src/app/[locale]/(pages)/accedi/__test__/page.test.tsx:1]: Sorry, I mean the eslint disable directive. Is it needed? I see no error if I remove it.

Comment (claudiodall) [reply src/app/[locale]/(pages)/accedi/__test__/page.test.tsx:1]: Ah, okay, you're right, I'll remove it

File: src/app/[locale]/_utils/__tests__/storage-utils.test.tsx (hunk 11 above)
Comment (claudiodall) [reply src/app/[locale]/_utils/__tests__/storage-utils.test.tsx:18]: You're right, let's create a more verbose description then. I also prefer using 'should' :)

File: src/app/[locale]/_utils/__tests__/common.test.tsx (hunk 8 above)
Comment (claudiodall) [reply src/app/[locale]/_utils/__tests__/common.test.tsx:10]: Sure

File: src/app/[locale]/_utils/__tests__/jwt.test.tsx (hunk 9 above)
Comment (claudiodall) [reply src/app/[locale]/_utils/__tests__/jwt.test.tsx:7]: Sure

File: src/app/[locale]/_utils/__tests__/jwt.test.tsx (hunk 10 above)
Comment (claudiodall) [reply src/app/[locale]/_utils/__tests__/jwt.test.tsx:22]: ok

File: src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx (hunk 6 above)
Comment (shadowsheep1) [reply src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx:70]: @claudiodall what about this comment?

Comment (claudiodall) [reply src/app/[locale]/(pages)/accedi/errore/__test__/errore.test.tsx:70]: I had missed the comment, sorry, fixed.

File: src/app/[locale]/(pages)/accedi/__test__/page.test.tsx [hunk 12]
```diff
@@ -1,25 +1,13 @@
...
 import Access from '../page';
+import * as it from '../../../../../dictionaries/it.json';
+import { renderWithProviders } from '@/app/[locale]/_utils/test-utils';
 
 describe('test suite for Access component', () => {
-  // eslint-disable-next-line @typescript-eslint/no-var-requires
-  const useRouter = jest.spyOn(require('next/router'), 'useRouter');
-  const locale = 'it';
+  test('should render Londing in case of ', async () => {
+    await renderWithProviders(<Access />);
```
Comment (shadowsheep1) [inline src/app/[locale]/(pages)/accedi/__test__/page.test.tsx:8]: Sorry, I don't understand the meaning of this test :). Does "Londing" stands for "Landing" or "Loading"? And should we render it in case of what? 🤔

Comment (shadowsheep1) [reply src/app/[locale]/(pages)/accedi/__test__/page.test.tsx:8]: Okay, sure. Just put it back with a more meaningful description in the future tests PRs 💪.

**Review by shadowsheep1 — APPROVED** [review]
LGTM 🚀
